compiler/qsc_circuit/src/circuit.rs

```diff
@@ -0,0 +1,612 @@
+                    let reg = (reg.q_id, reg.c_id);
+                    register_to_row.get(&reg)
+                })
+                .collect::<Vec<_>>();
+
+            // We'll need the entire range of rows for this operation so we can
+            // figure out the starting column and also so we can draw any
+            // vertical lines that cross wires.
+            let mut all_rows = targets.iter().chain(controls.iter()).collect::<Vec<_>>();
+            all_rows.sort_unstable();
+            let (first, last) = all_rows
+                .split_first()
+                .map_or((0, 0), |s| (***s.0, ***s.1.last().unwrap_or(s.0)));
+
+            // The starting column - the first available column in all
+            // the rows that this operation spans.
+            let mut column = 1;
+            for row in &rows[first..=last] {
+                if row.next_column > column {
+                    column = row.next_column;
+                }
+            }
+
+            // Add the operation to the diagram
+            for i in targets {
+                let row = &mut rows[*i];
+                if matches!(row.wire, Wire::Classical { .. }) && o.is_measurement {
+                    row.start_classical(column);
+                } else {
+                    row.add_gate(column, &o.gate, o.display_args.as_deref(), o.is_adjoint);
+                };
+            }
+
+            if o.is_controlled || o.is_measurement {
+                for i in controls {
+                    let row = &mut rows[*i];
+                    if matches!(row.wire, Wire::Qubit { .. }) && o.is_measurement {
+                        row.add_object(column, "M");
+                    } else {
+                        row.add_object(column, "●");
+                    };
+                }
+
+                // If we have a control wire, draw vertical lines spanning all
+                // control and target wires and crossing any in between
+                // (vertical lines may overlap if there are multiple controls/targets,
+                // this is ok in practice)
+                for row in rows.iter_mut().take(last + 1).skip(first) {
+                    row.add_vertical(column);
+                }
+            } else {
+                // No control wire. Draw dashed vertical lines to connect
+                // target wires if there are multiple targets
+                for row in rows.iter_mut().take(last + 1).skip(first) {
+                    row.add_dashed_vertical(column);
+                }
+            }
+        }
+
+        // Find the end column for the whole circuit so that
+        // all qubit wires will extend until the end
+        let end_column = rows
+            .iter()
+            .max_by_key(|r| r.next_column)
+            .map_or(1, |r| r.next_column);
+
+        // Draw the diagram
+        for row in rows {
+            row.fmt(f, end_column)?;
+        }
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
nitpick: same here, the tests should be moved to circuit/tests.rs to match the repo patterns.